File: ch1/py/h-structured.py
from langchain_openai import ChatOpenAI
from pydantic import BaseModel


class AnswerWithJustification(BaseModel):
    """An answer to the user's question along with justification for the answer."""

    answer: str
    """The answer to the user's question"""
    justification: str
    """Justification for the answer"""


llm = ChatOpenAI(model="gpt-3.5", temperature=0)
structured_llm = llm.with_structured_output(AnswerWithJustification)

response = structured_llm.invoke(
    "What weighs more, a pound of bricks or a pound of feathers")
print(response)

File: ch1/js/h-structured.js
import { ChatOpenAI } from '@langchain/openai';
import { z } from 'zod';

const answerSchema = z
  .object({
    answer: z.string().describe("The answer to the user's question"),
    justification: z.string().describe('Justification for the answer'),
  })
  .describe(
    "An answer to the user's question along with justification for the answer."
  );

const model = new ChatOpenAI({
  model: 'gpt-3.5-turbo',
  temperature: 0,
}).withStructuredOutput(answerSchema);

const response = await model.invoke(
  'What weighs more, a pound of bricks or a pound of feathers'
);
console.log(response);
